File: backend/data_utils/contracts.py
"""Load and validate the language-neutral pressure/posture contract."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any


PROJECT_ROOT = Path(__file__).resolve().parents[2]
CONTRACT_PATH = PROJECT_ROOT / "shared" / "contracts" / "posture.json"


class ContractError(ValueError):
    """Raised when a shared data contract is missing or inconsistent."""


def _read_contract(path: Path = CONTRACT_PATH) -> dict[str, Any]:
    try:
        document = json.loads(path.read_text(encoding="utf-8"))
    except FileNotFoundError as exc:
        raise ContractError(f"Shared posture contract was not found: {path}") from exc
    except json.JSONDecodeError as exc:
        raise ContractError(f"Shared posture contract is not valid JSON: {path}") from exc
    if not isinstance(document, dict):
        raise ContractError("Shared posture contract root must be a JSON object.")
    return document


def _positive_integer(value: Any, field: str) -> int:
    if not isinstance(value, int) or isinstance(value, bool) or value <= 0:
        raise ContractError(f"Contract field {field} must be a positive integer.")
    return value


def _build_contract_values(document: dict[str, Any]) -> dict[str, Any]:
    contract_version = document.get("contract_version")
    if not isinstance(contract_version, str) or not contract_version:
        raise ContractError("Contract field contract_version must be a non-empty string.")
    try:
        matrix = document["pressure_matrix"]
        posture_records = document["postures"]
        mirror_pairs = document["mirrored_action_pairs"]
    except KeyError as exc:
        raise ContractError(f"Shared posture contract is missing field {exc.args[0]!r}.") from exc

    if not isinstance(matrix, dict):
        raise ContractError("Contract field pressure_matrix must be an object.")
    matrix_shape = (
        _positive_integer(matrix.get("rows"), "pressure_matrix.rows"),
        _positive_integer(matrix.get("columns"), "pressure_matrix.columns"),
    )
    if matrix.get("index_order") != "row_column":
        raise ContractError("Only matrix[row][column] ordering is currently supported.")
    if not isinstance(posture_records, list) or not posture_records:
        raise ContractError("Contract field postures must be a non-empty array.")

    label_id_to_name: dict[int, str] = {}
    label_id_to_name_zh: dict[int, str] = {}
    action_to_label: dict[int, int] = {}
    mirrored_label: dict[int, int] = {}
    for record in posture_records:
        if not isinstance(record, dict):
            raise ContractError("Every posture entry must be an object.")
        label_id = record.get("id")
        name = record.get("key")
        name_zh = record.get("name_zh")
        actions = record.get("actions")
        mirrored_label_id = record.get("mirrored_label_id")
        if not isinstance(label_id, int) or isinstance(label_id, bool) or label_id < 0:
            raise ContractError("Every posture ID must be a non-negative integer.")
        if label_id in label_id_to_name:
            raise ContractError(f"Duplicate posture ID in contract: {label_id}")
        if not isinstance(name, str) or not name or not isinstance(name_zh, str) or not name_zh:
            raise ContractError(f"Posture {label_id} must have non-empty names.")
        if not isinstance(actions, list) or not actions:
            raise ContractError(f"Posture {label_id} must contain action IDs.")
        if not isinstance(mirrored_label_id, int) or isinstance(mirrored_label_id, bool):
            raise ContractError(f"Posture {label_id} has an invalid mirrored label ID.")

        label_id_to_name[label_id] = name
        label_id_to_name_zh[label_id] = name_zh
        mirrored_label[label_id] = mirrored_label_id
        for action in actions:
            if not isinstance(action, int) or isinstance(action, bool) or action < 0:
                raise ContractError(f"Posture {label_id} contains an invalid action ID.")
            if action in action_to_label:
                raise ContractError(f"Action {action} is assigned to multiple postures.")
            action_to_label[action] = label_id

    label_ids = set(label_id_to_name)
    if set(mirrored_label.values()) != label_ids:
        raise ContractError("Mirrored posture IDs must reference the defined posture IDs.")
    for label_id, mirror_id in mirrored_label.items():
        if mirrored_label.get(mirror_id) != label_id:
            raise ContractError("Mirrored posture mapping must be symmetric.")

    mirrored_action = {action: action for action in action_to_label}
    if not isinstance(mirror_pairs, list):
        raise ContractError("Contract field mirrored_action_pairs must be an array.")
    for pair in mirror_pairs:
        if not isinstance(pair, list) or len(pair) != 2:
            raise ContractError("Every mirrored action pair must contain two action IDs.")
        left, right = pair
        if left not in action_to_label or right not in action_to_label:
            raise ContractError("Mirrored action pair references an undefined action ID.")
        mirrored_action[left] = right
        mirrored_action[right] = left

    for action, mirror_action in mirrored_action.items():
        source_label = action_to_label[action]
        target_label = action_to_label[mirror_action]
        if target_label != mirrored_label[source_label]:
            raise ContractError(
                f"Mirrored action {action}->{mirror_action} conflicts with posture mapping."
            )

    return {
        "contract_version": contract_version,
        "matrix_shape": matrix_shape,
        "label_id_to_name": label_id_to_name,
        "label_id_to_name_zh": label_id_to_name_zh,
        "action_to_label": action_to_label,
        "mirrored_label": mirrored_label,
        "mirrored_action": mirrored_action,
    }


CONTRACT = _read_contract()
_VALUES = _build_contract_values(CONTRACT)
CONTRACT_VERSION: str = _VALUES["contract_version"]
MATRIX_SHAPE: tuple[int, int] = _VALUES["matrix_shape"]
LABEL_ID_TO_NAME: dict[int, str] = _VALUES["label_id_to_name"]
LABEL_ID_TO_NAME_ZH: dict[int, str] = _VALUES["label_id_to_name_zh"]
LABEL_NAME_TO_ID = {name: label_id for label_id, name in LABEL_ID_TO_NAME.items()}
ACTION_TO_LABEL: dict[int, int] = _VALUES["action_to_label"]
MIRRORED_LABEL: dict[int, int] = _VALUES["mirrored_label"]
MIRRORED_ACTION: dict[int, int] = _VALUES["mirrored_action"]


def action_to_label(action: int) -> int:
    """Return the static-posture label ID for an action from the contract."""

    try:
        return ACTION_TO_LABEL[action]
    except KeyError as exc:
        raise ValueError(
            f"Action {action} is not a static posture action "
            f"(expected one of {sorted(ACTION_TO_LABEL)})."
        ) from exc

File: frontend/src/core/contracts.ts
/**
 * 共享数据契约（shared/contracts/posture.json）的前端消费端。
 *
 * 架构原则：睡姿 ID、动作映射、镜像关系只以契约文档为唯一事实源。
 * - 模块初始化时以内置副本（src/core/contracts/posture.json）同步生效；
 * - 后端可用时通过 GET /api/contracts/posture 拉取远端契约并校验版本，
 *   版本一致才替换生效；不一致则保持本地基线并上报，UI 显示警告徽章。
 *
 * 校验规则与 backend/data_utils/contracts.py 对齐（语言中立契约）。
 */
import bundledRaw from './contracts/posture.json' with { type: 'json' };

export interface PostureContractEntry {
  id: number;
  key: string;
  name_zh: string;
  actions: number[];
  mirrored_label_id: number;
}

export interface PostureContract {
  contract_version: string;
  pressure_matrix: { rows: number; columns: number; index_order: string };
  postures: PostureContractEntry[];
  excluded_actions: Record<string, string>;
  mirrored_action_pairs: [number, number][];
}

export class ContractValidationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'ContractValidationError';
  }
}

function isPositiveInt(v: unknown, field: string): number {
  if (typeof v !== 'number' || !Number.isInteger(v) || v <= 0) {
    throw new ContractValidationError(`Contract field ${field} must be a positive integer.`);
  }
  return v;
}

/** 校验契约文档（与后端 _build_contract_values 对齐的规则） */
export function validateContract(raw: unknown): PostureContract {
  if (typeof raw !== 'object' || raw === null || Array.isArray(raw)) {
    throw new ContractValidationError('Contract root must be a JSON object.');
  }
  const doc = raw as Record<string, unknown>;
  const version = doc.contract_version;
  if (typeof version !== 'string' || !version) {
    throw new ContractValidationError('Contract field contract_version must be a non-empty string.');
  }

  const matrix = doc.pressure_matrix;
  if (typeof matrix !== 'object' || matrix === null || Array.isArray(matrix)) {
    throw new ContractValidationError('Contract field pressure_matrix must be an object.');
  }
  const m = matrix as Record<string, unknown>;
  const rows = isPositiveInt(m.rows, 'pressure_matrix.rows');
  const cols = isPositiveInt(m.columns, 'pressure_matrix.columns');
  if (m.index_order !== 'row_column') {
    throw new ContractValidationError('Only matrix[row][column] ordering is currently supported.');
  }

  const postures = doc.postures;
  if (!Array.isArray(postures) || postures.length === 0) {
    throw new ContractValidationError('Contract field postures must be a non-empty array.');
  }
  const seenIds = new Set<number>();
  const entries: PostureContractEntry[] = [];
  for (const record of postures) {
    if (typeof record !== 'object' || record === null) {
      throw new ContractValidationError('Every posture entry must be an object.');
    }
    const r = record as Record<string, unknown>;
    const id = r.id;
    if (typeof id !== 'number' || !Number.isInteger(id) || id < 0) {
      throw new ContractValidationError('Every posture ID must be a non-negative integer.');
    }
    if (seenIds.has(id)) throw new ContractValidationError(`Duplicate posture ID in contract: ${id}`);
    seenIds.add(id);
    const key = r.key;
    const nameZh = r.name_zh;
    if (typeof key !== 'string' || !key || typeof nameZh !== 'string' || !nameZh) {
      throw new ContractValidationError(`Posture ${id} must have non-empty names.`);
    }
    if (!Array.isArray(r.actions) || r.actions.length === 0) {
      throw new ContractValidationError(`Posture ${id} must contain action IDs.`);
    }
    if (typeof r.mirrored_label_id !== 'number' || !Number.isInteger(r.mirrored_label_id)) {
      throw new ContractValidationError(`Posture ${id} has an invalid mirrored label ID.`);
    }
    entries.push({
      id,
      key,
      name_zh: nameZh,
      actions: r.actions as number[],
      mirrored_label_id: r.mirrored_label_id,
    });
  }

  const excluded = doc.excluded_actions;
  const excludedOut: Record<string, string> = {};
  if (typeof excluded === 'object' && excluded !== null && !Array.isArray(excluded)) {
    for (const [k, v] of Object.entries(excluded as Record<string, unknown>)) {
      if (typeof v === 'string') excludedOut[k] = v;
    }
  }

  const pairs = doc.mirrored_action_pairs;
  const pairOut: [number, number][] = [];
  if (Array.isArray(pairs)) {
    for (const p of pairs) {
      if (
        Array.isArray(p) &&
        p.length === 2 &&
        p.every((n) => typeof n === 'number' && Number.isInteger(n))
      ) {
        pairOut.push([p[0], p[1]]);
      }
    }
  }

  return {
    contract_version: version,
    pressure_matrix: { rows, columns: cols, index_order: String(m.index_order) },
    postures: entries,
    excluded_actions: excludedOut,
    mirrored_action_pairs: pairOut,
  };
}

/** 内置契约（随构建打包，离线可用） */
export const LOCAL_CONTRACT: PostureContract = validateContract(bundledRaw);

let active: PostureContract = LOCAL_CONTRACT;

export interface ApplyResult {
  ok: boolean;
  error?: string;
  /** 远端版本与本地基线不一致 */
  versionMismatch?: boolean;
}

/**
 * 应用后端返回的契约文档。
 * 版本一致 → 替换生效；版本不一致 → 保持本地基线并标记 versionMismatch。
 */
export function applyRemoteContract(raw: unknown): ApplyResult {
  try {
    const parsed = validateContract(raw);
    if (parsed.contract_version !== LOCAL_CONTRACT.contract_version) {
      return {
        ok: false,
        versionMismatch: true,
        error: `远端契约 v${parsed.contract_version} 与前端基线 v${LOCAL_CONTRACT.contract_version} 不一致`,
      };
    }
    active = parsed;
    return { ok: true };
  } catch (e) {
    return { ok: false, error: e instanceof Error ? e.message : String(e) };
  }
}

export function contractVersion(): string {
  return active.contract_version;
}
export function matrixShape(): { rows: number; columns: number } {
  return { rows: active.pressure_matrix.rows, columns: active.pressure_matrix.columns };
}
export function isRemoteContractActive(): boolean {
  return active !== LOCAL_CONTRACT;
}

/** label id → 中文名（0 仰卧 / 1 俯卧 / 2 左侧卧 / 3 右侧卧） */
export function sleepPosNames(): Record<number, string> {
  const out: Record<number, string> = {};
  for (const p of active.postures) out[p.id] = p.name_zh;
  return out;
}

/** action → 睡姿 label id；excluded（0 空载 / 22 动态）返回 null */
export function actionToLabelId(action: number): number | null {
  if (Object.prototype.hasOwnProperty.call(active.excluded_actions, String(action))) return null;
  for (const p of active.postures) {
    if (p.actions.includes(action)) return p.id;
  }
  return null;
}

/** 镜像动作：契约镜像对内取对侧；无镜像（仰卧/俯卧等）返回 null */
export function mirrorAction(action: number): number | null {
  for (const [a, b] of active.mirrored_action_pairs) {
    if (action === a) return b;
    if (action === b) return a;
  }
  return null;
}
